### PR TITLE
allow IPCVal to be instantiated from config

### PR DESCRIPTION
**Pull request by joe-bowman**
Ref #2827 

(I have since seen #2847 which is a fix for the same issue; this PR has tests and docs too ;) )

* [x] Updated all relevant documentation in docs
* [x] Updated all code comments where relevant
* [x] Wrote tests
* [x] Updated CHANGELOG_PENDING.md

### CODE REVIEW

File: CHANGELOG_PENDING.md [hunk 1]
```diff
@@ -1,5 +1,7 @@
 # Pending
 
+* Add ability to instantiate IPCVal (ref #2827).
```
Comment (melekes) [inline CHANGELOG_PENDING.md]:
```suggestion
- [node] \#2827 add ability to instantiate IPCVal (@joe-bowman)
```

this should go under "IMPROVEMENTS"

File: docs/architecture/adr-008-priv-validator.md [hunk 2]
```diff
@@ -5,14 +5,17 @@ implementations:
...
 
-The SocketPV address can be provided via flags at the command line - doing so
-will cause Tendermint to ignore any "priv_validator.json" file and to listen on
-the given address for incoming connections from an external priv_validator
-process. It will halt any operation until at least one external process
-succesfully connected.
+Both TCPVal and IPCVal addresses can be provided via flags at the command line
+or in the configuration file; TCPVal addresses must be of the form
+`tcp://<ip_address>:<port>` and IPCVal addresses `unix:///path/to/file.sock` -
+doing so will cause Tendermint to ignore any "priv_validator.json" file.
```
Comment (melekes) [inline docs/architecture/adr-008-priv-validator.md]:
```suggestion
doing so will cause Tendermint to ignore any private validator files.
```

let's not focus on the concrete file extension (`.json`) and use "private validator file" or "node key file" instead.

File: node/node.go [hunk 3]
```diff
@@ -219,25 +219,50 @@ func NewNode(config *cfg.Config,
...
 
-		privValidator = pvsc
+			if err := pvsc.Start(); err != nil {
+				return nil, fmt.Errorf("Error starting private validator client: %v", err)
+			}
+
+			privValidator = pvsc
+
+		} else {
+			return nil, fmt.Errorf(
```
Comment (melekes) [inline node/node.go]:
```suggestion
			return nil, fmt.Errorf(
				"Error creating private validator: expected either tcp or unix protocols, got %s", 
				protocol)
```

File: node/node.go [hunk 4]
```diff
@@ -219,25 +219,50 @@ func NewNode(config *cfg.Config,
...
-			privKey = ed25519.GenPrivKey()
-			pvsc    = privval.NewTCPVal(
-				logger.With("module", "privval"),
-				config.PrivValidatorListenAddr,
-				privKey,
+		// If an address is provided, listen on the socket for a
+		// connection from an external signing process.
+		protocol, address := cmn.ProtocolAndAddress(config.PrivValidatorListenAddr)
+
+		if protocol == "unix" {
```
Comment (melekes) [inline node/node.go]: better to use a `switch` here

File: node/node.go [hunk 5]
```diff
@@ -219,25 +219,50 @@ func NewNode(config *cfg.Config,
...
 	if config.PrivValidatorListenAddr != "" {
-		var (
-			// TODO: persist this key so external signer
-			// can actually authenticate us
-			privKey = ed25519.GenPrivKey()
-			pvsc    = privval.NewTCPVal(
-				logger.With("module", "privval"),
-				config.PrivValidatorListenAddr,
-				privKey,
+		// If an address is provided, listen on the socket for a
```
Comment (melekes) [inline node/node.go]: can we extract this into a function?

File: node/node_test.go [hunk 6]
```diff
@@ -113,3 +118,111 @@ func TestNodeSetAppVersion(t *testing.T) {
...
+		ed25519.GenPrivKey(),
+	)
+	privval.RemoteSignerConnDeadline(5 * time.Millisecond)(rs)
+	privval.RemoteSignerConnRetries(1e6)(rs)
+
+	config := cfg.ResetTestRoot("node_priv_val_tcp_test")
+
+	config.SetPrivValidatorListenAddr(addr)
+
+  // kick off remote signer routine, and then start TM.
```
Comment (melekes) [inline node/node_test.go]: formatting is off

File: node/node_test.go [hunk 7]
```diff
@@ -113,3 +118,111 @@ func TestNodeSetAppVersion(t *testing.T) {
...
```
Comment (melekes) [inline node/node_test.go]: formatting is off

File: node/node_test.go [hunk 8]
```diff
@@ -113,3 +118,111 @@ func TestNodeSetAppVersion(t *testing.T) {
...
+	go func(rs *privval.RemoteSigner) {
+		defer rs.Stop()
```
Comment (melekes) [inline node/node_test.go]: defer usually goes after `Start()`, not before

File: node/node_test.go [hunk 9]
```diff
@@ -113,3 +118,111 @@ func TestNodeSetAppVersion(t *testing.T) {
...
+
+	n, err := DefaultNewNode(config, log.TestingLogger())
+
+	assert.NoError(t, err, "expected no err on DefaultNewNode")
+
+	assert.IsType(t, &privval.TCPVal{}, n.PrivValidator())
+}
+
+func TestNodeSetPrivValIPC(t *testing.T) {
+  tmpfile := "/tmp/kms." + cmn.RandStr(6) + ".sock"
```
Comment (melekes) [inline node/node_test.go]: formatting is off

File: config/config.go [hunk 10]
```diff
@@ -96,6 +96,11 @@ func (cfg *Config) SetRoot(root string) *Config {
 	return cfg
 }
 
+func (cfg *Config) SetPrivValidatorListenAddr(addr string) *Config {
```
Comment (melekes) [inline config/config.go]: I'd prefer not to have this function. It does not bring any value and encourages mutation.

File: node/node.go [hunk 11]
```diff
@@ -148,6 +148,48 @@ type Node struct {
...
+	default:
+		return nil, fmt.Errorf(
+			"Error creating private validator: expected either tcp or unix "+
+			"protocols, got %s",
+			protocol,
+		)
+	}
+
+	pvServ, ok := pvsc.(cmn.Service)
+	if !ok {
```
Comment (melekes) [inline node/node.go]: let's remove this check. It's clearly a developer's error and it's OK to panic if priv validator client does not implement Service.

File: config/config.go (hunk 10 above)
Comment (joe-bowman) [reply config/config.go]: Any pointers as to how I might override the priv_val_laddr config when testing option without it?

File: node/node.go (hunk 11 above)
Comment (joe-bowman) [reply node/node.go]: Sure thing.

**Comment by joe-bowman** [comment]
If 0.26.4 is planned to be release in preparation for GoS, is there any chance to get this merged in too? It doesn't touch any existing code, so should be low risk but will enable some participants to run their planned production configurations for GoS.

File: CHANGELOG_PENDING.md [hunk 12]
```diff
@@ -2,8 +2,6 @@
 
 ## v0.26.4
 
-*TBD*
```
[the remark this answers is not in the document]
Comment (melekes) [reply CHANGELOG_PENDING.md]:
```suggestion
*TBD*
```

**Pull request merged**